qemu/tests/migration_with_numa.py

```diff
@@ -0,0 +1,113 @@
+    :param params: Dictionary with the test parameters
+    :param env: Dictionary with test environment
+    """
+
+    def get_nodes_size(size_type='MemTotal', session=None):
+        """
+        Get the node size of each node in host/guest, descending sort with size
+
+        :param size_type: the type of the node size
+        :param session: ShellSession object
+
+        :return: a list of tuple include node id and node size
+        :rtype: list
+        """
+        numa_info = NumaInfo(session=session)
+        nodes_size = {}
+        numa_nodes = numa_info.online_nodes
+        for node in numa_nodes:
+            nodes_size[node] = float(numa_info.online_nodes_meminfo[node][size_type])
+        nodes_size = sorted(nodes_size.items(), key=lambda item: item[1], reverse=True)
+        return nodes_size
+
+    host_nodes_size = get_nodes_size(size_type='MemFree')
+    mem_devs = params.objects('mem_devs')
+    if len(host_nodes_size) < len(mem_devs):
+        test.cancel("Host do not have enough nodes for testing!")
+    for mem_dev in mem_devs:
+        size_mem = params.object_params(mem_dev).get('size_mem')
+        size_mem = float(normalize_data_size(size_mem, 'K'))
+        if host_nodes_size[0][1] >= int(size_mem):
+            params['host-nodes_mem_%s' % mem_dev] = str(host_nodes_size[0][0])
+            del host_nodes_size[0]
+        else:
+            test.cancel("host nodes do not have enough memory for testing!")
+
+    params['start_vm'] = 'yes'
+    env_process.preprocess(test, params, env)
+    vm = env.get_vm(params["main_vm"])
+    session = vm.wait_for_login()
+    # do migration
+    mig_timeout = float(params.get("mig_timeout", "3600"))
+    mig_protocol = params.get("migration_protocol", "tcp")
+    vm.migrate(mig_timeout, mig_protocol, env=env)
+    session = vm.wait_for_login()
+
+    os_type = params["os_type"]
+    if os_type == 'linux':
+        error_context.context("Check the memory in guest", logging.info)
+        # Use 30 plus the gap of 'MemTotal' in OS and '-m' in cli as threshold
+        mem_total = get_mem_info(session, 'MemTotal')
+        mem_total = float(normalize_data_size('%s KB' % mem_total))
+        error_context.context("MemTotal in guest os is %s MB"
+                              % mem_total, logging.info)
+        threshold = float(params.get_numeric("mem") - mem_total) + 30
+        error_context.context("The acceptable threshold is: %s"
+                              % threshold, logging.info)
+        guest_nodes_size = get_nodes_size(size_type='MemTotal', session=session)
+        guest_nodes_size = dict(guest_nodes_size)
+        for nodenr, node in enumerate(params.objects('guest_numa_nodes')):
+            mdev = params.get("numa_memdev_node%d" % nodenr)
+            if mdev:
+                mdev = mdev.split('-')[1]
+                size = float(normalize_data_size(params.get("size_mem_%s" % mdev)))
+                if size - guest_nodes_size[nodenr] > threshold:
```

Comment:
abs(size - guest_nodes_size[nodenr])

Reply:
I think it is almost impossible that the actual guest node size  is bigger than the size in qemu cli. But i am ok to add `abs`, thanks